### PR TITLE
feat(sql): add database/sql plugin

### PR DESCRIPTION
**Pull request by kagaya85**
Add go2sky plugin for `database/sql`

This PR is still WIP and may need sufficient tests before merge

Because this is a new plugin, we may need to discuss some implementation details 

Any comments or suggestions are welcome, thanks in advance

### CODE REVIEW

File: sql/stmt.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+	}
+	return s.Exec(values)
+}
+
+func (s *stmt) Query(args []driver.Value) (driver.Rows, error) {
+	return s.stmt.Query(args)
+}
+
+func (s *stmt) QueryContext(ctx context.Context, args []driver.NamedValue) (driver.Rows, error) {
+	span, err := s.tracer.CreateExitSpan(ctx, genOpName(s.dbType, "query"), s.addr, emptyInjectFunc)
```
Comment (arugal) [inline sql/stmt.go:78]: It looks like the [componentId](https://github.com/SkyAPM/go2sky/blob/master/span.go#L47) is missing. The `componentId` need to be defined in the [component-libraries.yml file](https://github.com/apache/skywalking/blob/master/oap-server/server-bootstrap/src/main/resources/component-libraries.yml), you can refer to the [Mysqli](https://github.com/apache/skywalking/blob/master/oap-server/server-bootstrap/src/main/resources/component-libraries.yml#L604) for PHP.

Comment (kagaya85) [reply sql/stmt.go:78]: Thanks for reminding this, I will add it, include `SetComponent` and `SetSpanLayer`

File: sql/common.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+		addr = re.FindString(dsn)
+		addr = addr[1 : len(addr)-1]
+	case IPV4:
+		// ipv4 addr
+		re := regexp.MustCompile(`((2(5[0-5]|[0-4]\d))|[0-1]?\d{1,2})(\.((2(5[0-5]|[0-4]\d))|[0-1]?\d{1,2})){3}:\d{1,5}`)
+		addr = re.FindString(dsn)
+	}
+	a.peer = addr
+	return
+}
```
Comment (kagaya85) [inline sql/common.go:81]: I think it is not reasonable for the plugin to parse the dsn to get the address of the database, even for the same database of different plugins, there may also be different dsn format, perhaps the user should directly configure the target address?

Comment (arugal) [reply sql/common.go:81]: The default is to parse the dsn, it also supports configuration via `AttributeOption` func, [like this](https://github.com/SkyAPM/go2sky/blob/master/trace.go#L47). What do you think :)

**Comment by arugal** [comment]
image

Hi @kagaya85 please update the license header :)

**Comment by kagaya85** [comment]
Now, the `Tx` struct contains a context of the transaction local span

if user use methods with context that contains span context, like `ExecContext`, new exit span will follow the span in the context, in addition, for methods without context(`Exec`, `Query`) or context do not contain span, new exit span will follow the local span in `Tx` struct

```go
type Tx struct {
	*sql.Tx

	db   *DB
	span go2sky.Span
	ctx  context.Context
}

func (tx *Tx) Exec(query string, args ...interface{}) (sql.Result, error) {
	return tx.ExecContext(tx.ctx, query, args)
}

func (tx *Tx) ExecContext(ctx context.Context, query string, args ...interface{}) (sql.Result, error) {
	if id := go2sky.SpanID(ctx); id == go2sky.EmptySpanID {
		// if ctx do not contain parent span, use transaction ctx instead
		ctx = tx.ctx
	}
	span, err := createSpan(ctx, tx.db.tracer, tx.db.opts, "execute")
	if err != nil {
		return nil, err
	}
	defer span.End()

	...

	return res, err
}
```

**Comment by kagaya85** [comment]
@arugal For the previous implementation, is it possible to keep it as a new plugin which based on `database/sql/driver`? 
this can lead to more detailed tracking and flexibility, such as using in some third-party sql libaray like sqlx

File: sql/tx.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+	st := tx.Tx.StmtContext(ctx, stmt.Stmt)
+	return &Stmt{
+		Stmt:  st,
+		db:    tx.db,
+		query: stmt.query,
+	}
+}
+
+func (tx *Tx) Exec(query string, args ...interface{}) (sql.Result, error) {
+	return tx.ExecContext(tx.ctx, query, args)
```
Comment (arugal) [inline sql/tx.go:85]:
```suggestion
	return tx.ExecContext(tx.ctx, query, args...)
```

fix `converting argument $1 type: unsupported type []interface {}, a slice of interface`.  @kagaya85  Please add a test case for `Tx.exec`

**Comment by arugal** [comment]
> @arugal For the previous implementation, is it possible to keep it as a new plugin which based on `database/sql/driver`?
> this can lead to more detailed tracking and flexibility, such as using in some third-party sql libaray like sqlx

 Good idea, we can do this in a new PR, after the current PR is merged

**Review by arugal — APPROVED** [review]
LGTM

## Local Test Results

### Database

image

### Trace
image

image

**Pull request merged**